### DISCUSSION TITLE
Using custom Transport Protocol (Cap'n Proto RPC)

### DISCUSSION THREAD

**Discussion by Tanguille** [libp2p/rust-libp2p#4420] [Q&A] [answered]
I want to use `libp2p` with `Cap'n Proto`. I can't seem to find any documentation in how you implement a protocol that isn't supported by `libp2p` so I hope someone here can help me.

I tried to compose a `multiaddr` with this custom format and it gave a compilation error:
```rust
swarm.listen_on("/ip4/0.0.0.0/capnp-rpc/0".parse()?)?;
```
[Here](https://github.com/Tanguille/sandbox) you can find my sandbox project where I'm testing this out if that would help to show what I mean.
Thanks in advance.

**Comment by thomaseizinger** [comment]
You seem to be conflating several things. I can recommend reading the general docs first: https://docs.libp2p.io/concepts/introduction/overview/

**Reply by Tanguille** [reply]
Hey, thanks for responding. I have read everything until PubSub. What part are you referring to that describes how you can implement a custom protocol?

**Reply by thomaseizinger** [reply]
Cool, thanks for reading through it! :)

As you know now, a _transport_ in libp2p is typically multiplexed, authenticated and encrypted. Captn proto provides none of these things.

But what I think you should be able to use is its serialization and data format. If you just want to do request-response style messages, then our `request_response` module might be of help.

If you want other message patterns, then you'd have to implement a custom `ConnectionHandler` and `NetworkBehaviour`. In that case though, I'd suggest to think about your problem differently. In rust-libp2p, we have modules that implement `NetworkBehaviour`. They are essentially plugins with a certain responsibility like managing a DHT, pub-sub, etc.

Captn Proto is a serialisation format. It doesn't really make sense IMO to ask: "How can I write a plugin for that?". You should be thinking about your usecase, which messages are to be exchanged and when and then perhaps use captn proto as an implementation detail for your serialisation, the same way you could just use cbor, json or anything else.

Hope that helps :)

**Reply by Tanguille** [reply]
This is certainly helpful, thanks. Might be that I totally misunderstood this but isn't RPC also a transport protocol? `Cap'n Proto` has an [RPC  implementation](https://capnproto.org/rpc.html) that has several benefits. That is what I wanna use and why I formulated my question that way. What I want is that the `Switch` makes a secure channel with multiplexing using this RPC. Is that something that would make sense to you?

**Reply by thomaseizinger** [reply, marked as answer]
I think you should be able to just use captn proto on top of an existing stream established by e.g. yamux layered on top of noise. Or a QUIC stream.

If I understand the above article correctly, captn proto just emits bytes that you still need to send over a socket. You can just send that over a stream instead :)
